substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,770 @@
+		NothingToSlash,
+		/// Failed to withdraw amount from Core Staking.
+		WithdrawFailed,
+		/// Operation not supported by this pallet.
+		NotSupported,
+	}
+
+	/// A reason for placing a hold on funds.
+	#[pallet::composite_enum]
+	pub enum HoldReason {
+		/// Funds held for stake delegation to another account.
+		#[codec(index = 0)]
+		Delegating,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Funds delegated by a delegator.
+		Delegated { agent: T::AccountId, delegator: T::AccountId, amount: BalanceOf<T> },
+		/// Funds released to a delegator.
+		Released { agent: T::AccountId, delegator: T::AccountId, amount: BalanceOf<T> },
+		/// Funds slashed from a delegator.
+		Slashed { agent: T::AccountId, delegator: T::AccountId, amount: BalanceOf<T> },
+	}
+
+	/// Map of Delegators to their `Delegation`.
+	///
+	/// Implementation note: We are not using a double map with `delegator` and `agent` account
+	/// as keys since we want to restrict delegators to delegate only to one account at a time.
+	#[pallet::storage]
+	pub(crate) type Delegators<T: Config> =
+		CountedStorageMap<_, Twox64Concat, T::AccountId, Delegation<T>, OptionQuery>;
+
+	/// Map of `Agent` to their `Ledger`.
+	#[pallet::storage]
+	pub(crate) type Agents<T: Config> =
+		CountedStorageMap<_, Twox64Concat, T::AccountId, AgentLedger<T>, OptionQuery>;
+
+	/// This pallet is not currently written with the intention of exposing any calls. But the
+	/// functions defined in the following impl block should act as a good reference for how the
+	/// exposed calls would look like when exposed.
+	impl<T: Config> Pallet<T> {
+		/// Register an account to become a stake `Agent`. Sometimes also called a `Delegatee`.
+		///
+		/// Delegators can authorize `Agent`s to stake on their behalf by delegating their funds to
+		/// them. The `Agent` can then use the delegated funds to stake to [`Config::CoreStaking`].
+		///
+		/// Implementation note: This function allows any account to become an agent. It is
+		/// important though that accounts that call [`StakingUnsafe::virtual_bond`] are keyless
+		/// accounts. This is not a problem for now since this is only used by other pallets in the
+		/// runtime which use keyless account as agents. If we later want to expose this as a
+		/// dispatchable call, we should derive a sub-account from the caller and use that as the
+		/// agent account.
+		pub fn register_agent(
+			origin: OriginFor<T>,
+			reward_account: T::AccountId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			// Existing `agent` cannot register again.
+			ensure!(!Self::is_agent(&who), Error::<T>::NotAllowed);
+
+			// A delegator cannot become an `agent`.
+			ensure!(!Self::is_delegator(&who), Error::<T>::NotAllowed);
+
+			// They cannot be already a direct staker in the staking pallet.
+			ensure!(Self::not_direct_staker(&who), Error::<T>::AlreadyStaking);
```

Comment:
```suggestion
			ensure!(!Self::direct_staker(&who), Error::<T>::AlreadyStaking);
```

In general I think it's simpler to avoid negations in the method names. e.g. `!not_direct_staker` becomes double negation and it's harder to read.